dvc/cache/local.py

```diff
@@ -154,14 +155,30 @@ def unprotect(self, path_info):
             self._unprotect_file(path_info)
 
     def protect(self, path_info):
-        self.tree.chmod(path_info, self.CACHE_MODE)
+        try:
+            os.chmod(path_info, self.CACHE_MODE)
+        except OSError:
+            # NOTE: not being able to protect cache file is not fatal, it
+            # might happen on funky filesystems (e.g. Samba, see #5255),
+            # read-only filesystems or in a shared cache scenario.
+            logger.debug("failed to protect '%s'", path_info, exc_info=True)
 
     def is_protected(self, path_info):
-        import stat
-
         try:
             mode = os.stat(path_info).st_mode
         except FileNotFoundError:
             return False
 
         return stat.S_IMODE(mode) == self.CACHE_MODE
+
+    def set_exec(self, path_info):
```

Comment:
`set_exec` really belongs to `checkout`(e.g. it should be ignored for hardlinks/symlinks), but for now it is out of scope. It got affected in this PR because it was using the same chmod logic.